timer/timer_test.go

```diff
@@ -0,0 +1,435 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package timer_test
+
+import (
+	"github.com/DataDog/go-libddwaf/v2/internal/unsafe"
+	"github.com/DataDog/go-libddwaf/v2/timer"
+	"github.com/stretchr/testify/require"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+)
+
+func hasExpired(t *testing.T, timer timer.Timer, duration time.Duration) {
+	require.Greater(t, timer.Spent(), duration)
+	require.Zero(t, timer.Remaining())
+	require.True(t, timer.Expired())
+}
+
+func hasSumExpired(t *testing.T, timer timer.NodeTimer, duration time.Duration) {
+	require.Greater(t, timer.SumSpent(), duration)
+	require.Zero(t, timer.SumRemaining())
+	require.True(t, timer.SumExpired())
+}
+
+func TestBasic(t *testing.T) {
+	t.Run("start-is-not-blank", func(t *testing.T) {
+		newTimer, err := timer.NewTimer(timer.WithBudget(time.Second))
+		require.NoError(t, err)
+		require.NotEqual(t, time.Time{}, newTimer.Start())
+	})
+
+	t.Run("stop-means-stop", func(t *testing.T) {
+		newTimer, err := timer.NewTimer(timer.WithBudget(time.Second))
+		require.NoError(t, err)
+		newTimer.Start()
+		time.Sleep(1 * time.Millisecond)
+		spent := newTimer.Stop()
+		require.Equal(t, newTimer.Spent(), spent)
+
+		// The timer was indeed stopped
+		time.Sleep(1 * time.Millisecond)
+		require.Equal(t, newTimer.Spent(), spent)
+	})
+
+	t.Run("timed=spent", func(t *testing.T) {
+		newTimer, err := timer.NewTimer(timer.WithBudget(time.Second))
+		require.NoError(t, err)
+		spent := newTimer.Timed(func(timer timer.Timer) {
+			time.Sleep(1 * time.Millisecond)
+		})
+
+		require.GreaterOrEqual(t, newTimer.Spent(), time.Millisecond)
+		require.Equal(t, newTimer.Spent(), spent)
+	})
+
+	t.Run("stop=spent", func(t *testing.T) {
+		newTimer, err := timer.NewTimer(timer.WithBudget(time.Second))
+		require.NoError(t, err)
+		newTimer.Start()
+		time.Sleep(1 * time.Millisecond)
+		spent := newTimer.Stop()
+
+		require.GreaterOrEqual(t, newTimer.Spent(), time.Millisecond)
+		require.Equal(t, newTimer.Spent(), spent)
+	})
+}
+
+func TestSum(t *testing.T) {
+
+	if os.Getenv("GITHUB_RUN_ID") != "" {
+		t.Skip("Skipping test in CI environment")
+	}
```

Comment:
... because ...?

I'll note that the `timeCache` / `clock` thing could be an interface that exposes `now() time.Time`, which you could then have a "real" implementation and a "fake" implementation that you advance "manually" for tests like these.

Reply:
> ... because ...?
Because I still have to test the real implementation at some point even if with most tests I have ways for the tests to not be flaky.

Yes but I did not create a fake clock and fake timer for now because I don't want the PR to be bigger than this. And if I create a fake clock / timer I would also have to test it carefully before using it to test other stuff.

Reply:
The spirit of that:

>... because ...?

Was to ask you to explain WHY you're skipping this in the CI environment in a comment or within the skip message, so it's easier for the next engineer around there to determine if this is still relevant or not ;)

Reply:
Reworked the test to make it work on the CI